spring-cloud-aws-cognito/src/main/java/io/awspring/cloud/cognito/CognitoTemplate.java

```diff
@@ -0,0 +1,129 @@
+			String userPoolId, String clientSecret) {
+		Assert.notNull(cognitoIdentityProviderClient, "cognitoIdentityProviderClient is required");
+		Assert.notNull(clientId, "clientId is required");
+		Assert.notNull(userPoolId, "userPoolId is required");
+		this.cognitoIdentityProviderClient = cognitoIdentityProviderClient;
+		this.clientId = clientId;
+		this.userPoolId = userPoolId;
+		this.clientSecret = clientSecret;
+	}
+
+	@Override
+	public AdminInitiateAuthResponse login(String username, String password) {
+		AdminInitiateAuthRequest adminInitiateAuthRequest = AdminInitiateAuthRequest.builder().userPoolId(userPoolId)
+				.clientId(clientId).authFlow(AuthFlowType.ADMIN_USER_PASSWORD_AUTH)
+				.authParameters(resolveAuthParameters(username, password)).build();
+		return cognitoIdentityProviderClient.adminInitiateAuth(adminInitiateAuthRequest);
+	}
+
+	@Override
+	public AdminCreateUserResponse createUser(String username, String password, List<AttributeType> attributeTypes) {
+		AdminCreateUserRequest createUserRequest = AdminCreateUserRequest.builder().userPoolId(userPoolId)
+				.username(username).temporaryPassword(password).userAttributes(attributeTypes).build();
+		return cognitoIdentityProviderClient.adminCreateUser(createUserRequest);
+	}
+
+	@Override
+	public ForgotPasswordResponse resetPassword(String username) {
+		ForgotPasswordRequest.Builder forgotPasswordRequestBuilder = ForgotPasswordRequest.builder().clientId(clientId)
+				.username(username);
+		if (this.clientSecret != null) {
+			forgotPasswordRequestBuilder.secretHash(CognitoUtils.calculateSecretHash(clientId, clientSecret, username));
+		}
+		ForgotPasswordRequest forgotPasswordRequest = forgotPasswordRequestBuilder.build();
+
+		return cognitoIdentityProviderClient.forgotPassword(forgotPasswordRequest);
+	}
+
+	@Override
+	public ConfirmForgotPasswordResponse confirmResetPassword(String username, String confirmationCode,
+			String newPassword) {
+		ConfirmForgotPasswordRequest confirmForgotPasswordRequest = ConfirmForgotPasswordRequest.builder()
+				.clientId(clientId).username(username).password(newPassword).confirmationCode(confirmationCode)
+				.secretHash(CognitoUtils.calculateSecretHash(clientId, clientSecret, username)).build();
+		return cognitoIdentityProviderClient.confirmForgotPassword(confirmForgotPasswordRequest);
+	}
+
+	@Override
+	public RespondToAuthChallengeResponse setPermanentPassword(String session, String username, String password) {
+		RespondToAuthChallengeRequest respondToAuthChallengeRequest = RespondToAuthChallengeRequest.builder()
+				.clientId(clientId).challengeName(ChallengeNameType.NEW_PASSWORD_REQUIRED).session(session)
+				.challengeResponses(Map.of(CognitoParameters.USERNAME_PARAM_NAME, username,
+						CognitoParameters.NEW_PASSWORD_PARAM_NAME, password, CognitoParameters.SECRET_HASH_PARAM_NAME,
+						CognitoUtils.calculateSecretHash(clientId, clientSecret, username)))
```

Comment:
if `clientSecret` can be null, this may cause null pointer exception. Is `clientSecret` mean to be nullable?

Reply:
Yes, indeed. It might be null for public clients (by design, they don't have a `clientSecret`), and those clients mainly use SPA apps. However public clients might be configured to allow auth operations with server-side auth credentials flow. Will make these flows null-safe